File: relay-service/app/realtime_broadcast.py
import logging

import httpx

logger = logging.getLogger(__name__)


def broadcast(supabase_url: str, service_key: str, topic: str, event: str, payload: dict) -> None:
    """Send a Realtime broadcast message via Supabase's server-side REST
    endpoint (POST /realtime/v1/api/broadcast), rather than supabase-py's
    Client.channel(): the installed supabase-py/realtime version's
    synchronous channel (SyncRealtimeClient/SyncRealtimeChannel) has no
    connect()/subscribe()/send() implemented yet, only the async client does.
    This REST endpoint delivers to the same channel/topic either way.
    """
    url = f"{supabase_url}/realtime/v1/api/broadcast"
    headers = {
        "apikey": service_key,
        "Authorization": f"Bearer {service_key}",
        "Content-Type": "application/json",
    }
    body = {"messages": [{"topic": topic, "event": event, "payload": payload}]}
    response = httpx.post(url, headers=headers, json=body, timeout=5.0)
    response.raise_for_status()

File: supabase/functions/_shared/realtimeBroadcast.ts
// Mirrors relay-service/app/realtime_broadcast.py — same server-side REST
// endpoint (POST /realtime/v1/api/broadcast) used for the existing
// "price-ticks" channel, reused here for a "transactions" channel so the
// frontend can toast every logged transaction (voice-initiated or
// cron-fired recurring rules) without needing table-level Realtime/RLS
// access.
export async function broadcastRealtime(
  supabaseUrl: string,
  serviceKey: string,
  topic: string,
  event: string,
  payload: Record<string, unknown>,
): Promise<void> {
  try {
    await fetch(`${supabaseUrl}/realtime/v1/api/broadcast`, {
      method: "POST",
      headers: {
        apikey: serviceKey,
        Authorization: `Bearer ${serviceKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ messages: [{ topic, event, payload }] }),
    });
  } catch {
    // Best-effort — a missed toast shouldn't fail the underlying transaction.
  }
}
